### PR TITLE
COMPAT: Add keepdims and friends to validation

### PR DESCRIPTION
**Pull request by TomAugspurger**
xref https://github.com/pandas-dev/pandas/pull/24227

### CODE REVIEW

File: pandas/compat/numpy/function.py [hunk 1]
```diff
@@ -225,16 +226,32 @@ def validate_cum_func_with_skipna(skipna, args, kwargs, name):
 STAT_FUNC_DEFAULTS = OrderedDict()
 STAT_FUNC_DEFAULTS['dtype'] = None
 STAT_FUNC_DEFAULTS['out'] = None
+
+PROD_DEFAULTS = SUM_DEFAULTS = STAT_FUNC_DEFAULTS.copy()
```
Comment (TomAugspurger) [inline pandas/compat/numpy/function.py]: This is a bit complicated to get the order of the keywords correct.

File: pandas/compat/numpy/function.py [hunk 2]
```diff
@@ -225,16 +226,32 @@ def validate_cum_func_with_skipna(skipna, args, kwargs, name):
...
+MEDIAN_DEFAULTS = STAT_FUNC_DEFAULTS.copy()
+MEDIAN_DEFAULTS['overwrite_input'] = False
+MEDIAN_DEFAULTS['keepdims'] = False
+
+STAT_FUNC_DEFAULTS['keepdims'] = False
+
 validate_stat_func = CompatValidator(STAT_FUNC_DEFAULTS,
                                      method='kwargs')
-validate_sum = CompatValidator(STAT_FUNC_DEFAULTS, fname='sort',
+validate_sum = CompatValidator(SUM_DEFAULTS, fname='sum',
```
Comment (TomAugspurger) [inline pandas/compat/numpy/function.py]: I think the incorrect `fname` was a copy-paste error.

File: pandas/tests/series/test_analytics.py [hunk 3]
```diff
@@ -1641,6 +1641,37 @@ def test_value_counts_categorical_not_ordered(self):
         tm.assert_series_equal(s.value_counts(normalize=True), exp)
         tm.assert_series_equal(idx.value_counts(normalize=True), exp)
 
+    @pytest.mark.parametrize("func", [np.any, np.all])
```
Comment (TomAugspurger) [inline pandas/tests/series/test_analytics.py]: @gfyoung are these tests for uses of `nv.validate*` collected anywhere?  I couldn't find where they were.

File: pandas/compat/numpy/function.py (hunk 2 above)
Comment (gfyoung) [reply pandas/compat/numpy/function.py]: Good catch 🙂

File: pandas/tests/series/test_analytics.py (hunk 3 above)
Comment (gfyoung) [reply pandas/tests/series/test_analytics.py]: No, we do not.  Leaving them organized as they are (in their respective function tests) is fine.

File: pandas/core/generic.py [hunk 4]
```diff
@@ -10834,7 +10834,12 @@ def _make_min_count_stat_function(cls, name, name1, name2, axis_descr, desc,
     def stat_func(self, axis=None, skipna=None, level=None, numeric_only=None,
                   min_count=0,
                   **kwargs):
-        nv.validate_stat_func(tuple(), kwargs, fname=name)
+        if name == 'sum':
+            nv.validate_sum(tuple(), kwargs)
+        elif name == 'prod':
+            nv.validate_prod(tuple(), kwargs)
```
Comment (jreback) [inline pandas/core/generic.py]: can we make this more generic? IOW have ``nv.validate_statu_func`` dispatch based on fname?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: I think that would make sense, but it isn't really written that way right now. Right now we just have instances of CompatValidator sitting in `function.py`, and we choose the right one to call. 

It'd be nice to have a decorator that did it for us

```python
@validate_numpy
def mean(self, ...):
    ...
```

then we don't have the duplication of the function name. But I think that's a decent sized refactor of how things are done now.

Comment (gfyoung) [reply pandas/core/generic.py]: > But I think that's a decent sized refactor of how things are done now.

Agreed.  At the time, I wrote it that way to keep things explicit, and it made it slightly easier to handle the details of each analogous `numpy` function.

That being said, I think refactoring to dispatching seems reasonable as well but would be best served for investigation and execution in a follow-up.

**Review by gfyoung — APPROVED** [review]

**Review by jreback — APPROVED** [review]

File: pandas/core/generic.py (hunk 4 above)
Comment (jreback) [reply pandas/core/generic.py]: ok for now. let's followup and see if can reduce need to specify names like this.

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**